scheme/parsec-tpm/evidence_handler.go

```diff
@@ -104,3 +225,185 @@ func parsecTpmLookupKey(scope, tenantID, class, instance string) string {
+
+	// once the signature on the token is verified, we can claim the HW is
+	// authentic
+	appraisal.TrustVector.Hardware = ear.GenuineHardwareClaim
+
+	ev, err := mapToEvidence(evidence)
+	if err != nil {
+		return fmt.Errorf("failed attestation results: %w", err)
+	}
+	if ev.Pat == nil {
+		return errors.New("no platform token to evaluate")
+	}
+
+	attInfo, err := ev.Pat.GetAttestationInfo()
+	if err != nil {
+		return fmt.Errorf("unable to get attestation information: %w", err)
+	}
+	if attInfo == nil {
+		return errors.New("attestation information in platform token is nil")
+	}
+
+	pcrs := attInfo.PCR.PCRinfo.PCRs
+	hashAlgID := attInfo.PCR.PCRinfo.HashAlgID
+	pcrDigest := attInfo.PCR.PCRDigest
+
+	// Match the Evidence PCR against the endorsements to generate a subset
+	// of matching endorsements
+	eds, err := matchPCRs(pcrs, hashAlgID, endorsements)
+	if err != nil {
+		appraisal.TrustVector.Executables = ear.UnrecognizedRuntimeClaim
+		log.Error("\n match PCR failed")
+		return fmt.Errorf("match PCR failed: %w", err)
+	}
+
+	if err := matchPCRDigest(pcrDigest, hashAlgID, eds); err != nil {
+		appraisal.TrustVector.Executables = ear.UnrecognizedRuntimeClaim
+		log.Error("\n match PCR Digest failed")
+		return fmt.Errorf("match failed for PCR Digest: %w", err)
+	}
+
+	appraisal.TrustVector.Executables = ear.ApprovedRuntimeClaim
+	appraisal.TrustVector.InstanceIdentity = ear.TrustworthyInstanceClaim
+	log.Debug("\n matchPCRs and matchPCRDigest Success")
+
+	// Populate Veraison Key Attesation Extension
+	if ev.Kat == nil {
+		return errors.New("no key attestation information")
+	}
+	key, err := ev.Kat.DecodePubArea()
+	if err != nil {
+		return fmt.Errorf("decoding failed for Public Key: %w", err)
+	}
+	kd, err := x509.MarshalPKIXPublicKey(key)
+	if err != nil {
+		return fmt.Errorf("unable to marshal public key: %w", err)
+	}
+	pubkey := base64.StdEncoding.EncodeToString(kd)
+	appraisal.AppraisalExtensions.VeraisonKeyAttestation = &map[string]interface{}{
+		"akpub": pubkey,
+	}
+	appraisal.UpdateStatusFromTrustVector()
+	appraisal.VeraisonAnnotatedEvidence = &evidence
+	return nil
+}
+
+func mapToEvidence(in map[string]interface{}) (*parsectpm.Evidence, error) {
+	evidence := &parsectpm.Evidence{}
+	data, err := json.Marshal(in)
+	if err != nil {
+		return nil, err
+	}
+	err = evidence.FromJSON(data)
+	if err != nil {
+		return nil, fmt.Errorf("unable to map to evidence: %w", err)
+	}
+	return evidence, err
+}
+
+// match the evidence PCR's against the received Endorsements
+func matchPCRs(pcrs []int, algID uint64, endorsements []Endorsements) ([]Endorsements, error) {
+	var eds []Endorsements
+
+	if len(pcrs) == 0 {
+		return nil, errors.New("no evidence pcrs to match")
+	}
+
+	// Sort the PCRs first
+	sort.Ints(pcrs)
+	for i, pcr := range pcrs {
+		matched := false
+		for _, end := range endorsements {
+			if (end.Attr.PCR == nil) || (end.Attr.AlgID == nil) {
+				return nil, errors.New("no valid endorsements to match")
```

Comment:
Not sure I understand the error message here? This happens if a particular endorsement has nil fields. We don't know that there aren't any valid endorsements further in the `endorsements`; perhaps this should say something like `"malformed endorsement: {end}"` or something?